core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -929,32 +938,51 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       inode = inodePath.getInode();
       auditContext.setSrcInode(inode);
 
-      List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
-        try {
-          mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
+      List<FileInfo> ret = listStatusInternal(inodePath, auditContext,
+          listStatusOptions.isRecursive(), inodePath);
+
+      auditContext.setSucceeded(true);
+      Metrics.FILE_INFOS_GOT.inc();
+      return ret;
+    }
+  }
+
+  private List<FileInfo> listStatusInternal(LockedInodePath currInodePath,
+                                                  AuditContext auditContext,
+                                                  boolean recursive,
+                                                  LockedInodePath originalInodePath)
+      throws FileDoesNotExistException, UnavailableException, AccessControlException {
+    List<FileInfo> statusList = new ArrayList<>();
+    Inode<?> inode = currInodePath.getInode();
+    if (inode.isDirectory()) {
+      try {
+        mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, currInodePath);
+      } catch (AccessControlException e) {
+        auditContext.setAllowed(false);
+        if (recursive) {
+          return statusList;
+        } else {
           throw e;
         }
+      }
+      if (recursive || currInodePath.equals(originalInodePath)) {
         for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-          child.lockReadAndCheckParent(inode);
-          try {
-            // the path to child for getPath should already be locked.
-            tempInodePath.setDescendant(child, mInodeTree.getPath(child));
-            ret.add(getFileInfoInternal(tempInodePath));
-          } finally {
-            child.unlockRead();
+          try (LockedInodePath childInodePath = mInodeTree.lockFullInodePath(child.getId(),
```

Comment:
hrmmm, is this because we don't have a temp way to extend a lock list?

Reply:
Yes. For reentrant locks, this is fine though, since we will never upgrade a lock from read to write?